core/desugarSwitchFuns.ml

```diff
@@ -0,0 +1,116 @@
+open Sugartypes
+open Utility
+open SourceCode
+
+(* This module desugars pattern-matching functions
+
+  This transformation convert `switch` functions of the form:
+
+  fun foo(a1, ..., an) switch {
+    case (p1_1, ..., p1_n) -> b_1
+    ...
+    case (pm_1, pm_n) -> b_m
+  }
+
+  to standard functions of the form:
+
+  fun foo(a1 as x1, ..., an as xn) {
+    switch ((x1, ..., xn)) {
+      case (p1_1, ..., p1_n) -> b_1
+      ...
+      case (pm_1, ..., pm_n) -> b_m
+      case (_, ..., _) -> error("non-exhaustive")
+  }
+
+  The last non-exhaustive case with wild card pattern is always attached to the end of switch body.
+
+*)
+
+let with_pos = SourceCode.WithPos.make
+
+let pattern_matching_sugar =
+  Settings.(
+    flag "pattern_matching_sugar"
+    |> synopsis
+         "Toggles whether to enable the switch pattern matching syntax sugar"
+    |> convert parse_bool
+    |> sync)
+
+let pattern_matching_sugar_guard pos =
+  let pattern_matching_sugar_disabled pos =
+    Errors.disabled_extension ~pos ~setting:("pattern_matching_sugar", true) "Pattern Matching Sugar"
+  in
+  if not (Settings.get pattern_matching_sugar)
+  then raise (pattern_matching_sugar_disabled pos)
+
+let nullary_guard pss pos =
+  let nullary_error pos =
+    Errors.desugaring_error ~pos:pos ~stage:Errors.DesugarSwitchFuns ~message:"Can't match over nullary function"
+  in
+  match pss with
+    | [] -> raise (nullary_error pos)
+    | _ -> ()
+
+let switch_fun_currying_guard pos args =
+  match args with
+  | [arg] -> arg
+  | _ -> raise (Errors.Type_error (pos, "Curried switch functions are not yet supported."))
+
+let construct_normal_funlit funlit_pos patterns cases =
+  pattern_matching_sugar_guard funlit_pos;
+  let patterns = switch_fun_currying_guard funlit_pos patterns in
+  nullary_guard patterns funlit_pos;
+  (* bind the arguments with unique var name *)
+  let pat_first_pos = WithPos.pos (List.nth patterns 0) in
+  let name_list = List.map (fun pat -> (pat, Utility.gensym())) patterns in
+  let switch_tuple = List.map (fun (_, name) -> with_pos (Var name)) name_list in
+  (* assemble exhaustive handler *)
+  let exhaustive_patterns = with_pos (Pattern.Any) in
```

Comment:
Note: This pattern/case may be redundant, though, currently we do not check for redundancy or exhaustiveness of patterns.